### PR TITLE
bpf: optimize dsr and add ipip support for lb-only

### PR DESCRIPTION
**Pull request by borkmann**
See commit msgs for details.

### CODE REVIEW

**Comment by borkmann** [comment]
test-me-please

File: bpf/lib/nodeport.h [hunk 1]
```diff
@@ -576,61 +575,60 @@ static __always_inline int nodeport_lb6(struct __ctx_buff *ctx,
...
 	if (!revalidate_data(ctx, &data, &data_end, &ip6))
 		return DROP_INVALID;
 
 	backend_local = lookup_ip6_endpoint(ip6);
 	if (!backend_local && lb6_svc_is_hostport(svc))
 		return DROP_INVALID;
 
-	switch (ret) {
-	case CT_NEW:
+	if (backend_local || (!backend_local && !nodeport_uses_dsr6(&tuple))) {
```
Comment (pchaigno) [inline bpf/lib/nodeport.h]:
```suggestion
	if (backend_local || !nodeport_uses_dsr6(&tuple)) {
```
Maybe also worth adding a small comment with basically what's in the commit message?

File: bpf/lib/nodeport.h [hunk 2]
```diff
@@ -1293,67 +1290,66 @@ static __always_inline int nodeport_lb4(struct __ctx_buff *ctx,
...
 	if (!revalidate_data(ctx, &data, &data_end, &ip4))
 		return DROP_INVALID;
 
 	backend_local = lookup_ip4_endpoint(ip4);
 	if (!backend_local && lb4_svc_is_hostport(svc))
 		return DROP_INVALID;
 
-	switch (ret) {
-	case CT_NEW:
+	if (backend_local || (!backend_local && !nodeport_uses_dsr4(&tuple))) {
```
Comment (pchaigno) [inline bpf/lib/nodeport.h]: Ditto.

File: bpf/lib/nodeport.h [hunk 3]
```diff
@@ -331,22 +364,31 @@ int tail_nodeport_ipv6_dsr(struct __ctx_buff *ctx)
...
 
 	if (!revalidate_data(ctx, &data, &data_end, &ip6))
 		return DROP_INVALID;
-
-	dport   = ctx_load_meta(ctx, CB_SVC_PORT);
+#if DSR_ENCAP_MODE == DSR_ENCAP_IPIP
+	addr.p1 = ctx_load_meta(ctx, CB_BE_ADDR_V6_1);
+	addr.p2 = ctx_load_meta(ctx, CB_BE_ADDR_V6_2);
+	addr.p3 = ctx_load_meta(ctx, CB_BE_ADDR_V6_3);
+	addr.p4 = ctx_load_meta(ctx, CB_BE_ADDR_V6_4);
```
Comment (pchaigno) [inline bpf/lib/nodeport.h]: What does BE stand for? Backend?

Why do we need constants distinct from `CB_SVC_ADDR_V6_X`? If we used the same, we could reduce code duplication here.

File: bpf/lib/nodeport.h [hunk 4]
```diff
@@ -331,22 +364,31 @@ int tail_nodeport_ipv6_dsr(struct __ctx_buff *ctx)
...
 
-	ret = set_dsr_ext6(ctx, ip6, &addr, dport);
+	ret = dsr_set_ext6(ctx, ip6, &addr,
+			   ctx_load_meta(ctx, CB_SVC_PORT));
+#else
+# error "Invalid load balancer DSR encapsulation mode!"
+#endif
 	if (ret)
-		return ret;
+		return DROP_INVALID;
```
Comment (pchaigno) [inline bpf/lib/nodeport.h]: Why is the error code returned by `dsr_set_xxx` overwritten?

File: bpf/lib/nodeport.h [hunk 5]
```diff
@@ -26,6 +26,10 @@
 #ifndef NATIVE_DEV_IFINDEX
 #define NATIVE_DEV_IFINDEX 0
 #endif
+#ifndef DSR_ENCAP_MODE
+#define DSR_ENCAP_MODE 0
+#define DSR_ENCAP_IPIP 2
```
Comment (pchaigno) [inline bpf/lib/nodeport.h:31]: We should probably also compile-test this in `bpf/Makefile`.

File: bpf/lib/nodeport.h (hunk 3 above)
Comment (borkmann) [reply bpf/lib/nodeport.h]: BE for backend, but yeah, I think I'll consolidate both instead.

File: bpf/lib/nodeport.h (hunk 4 above)
Comment (borkmann) [reply bpf/lib/nodeport.h]: I've adapted the same for v4/v6, one of the two was just doing DROP_INVALID already. My thinking was to switch this over as well so it's consistent, I can also just pass through ret in both cases, iirc nothing else aside from DROP_INVALID would be returned currently.

Comment (pchaigno) [reply bpf/lib/nodeport.h]: `dsr_set_ipip6` may also return `DROP_WRITE_ERROR` (admittedly, both error codes are close :slightly_smiling_face:).

**Comment by borkmann** [comment]
test-me-please

File: Documentation/cmdref/cilium-agent.md [hunk 6]
```diff
@@ -35,6 +35,7 @@ cilium-agent [flags]
       --bpf-fragments-map-max int                            Maximum number of entries in fragments tracking map (default 8192)
       --bpf-lb-acceleration string                           BPF load balancing acceleration via XDP ("native", "disabled") (default "disabled")
       --bpf-lb-algorithm string                              BPF load balancing algorithm ("random", "maglev") (default "random")
+      --bpf-lb-dsr-dispatch string                           BPF load balancing DSR dispatch method ("opt", "ipip") (default "opt")
```
Comment (brb) [inline Documentation/cmdref/cilium-agent.md:38]: Maybe we should extend by saying `"ipip" (only in lb-mode with XDP)`?

File: daemon/cmd/kube_proxy_replacement.go [hunk 7]
```diff
@@ -270,6 +279,16 @@ func initKubeProxyReplacementOptions() (strict bool) {
 					option.NodePortAcceleration, option.NodePortAccelerationDisabled, option.TunnelName, option.TunnelDisabled)
 			}
 		}
+
+		if option.Config.NodePortMode == option.NodePortModeDSR &&
+			option.Config.LoadBalancerDSRDispatch == option.DSRDispatchIPIP {
+			if option.Config.DatapathMode != datapathOption.DatapathModeLBOnly {
+				log.Fatalf("DSR dispatch mode %s only supported for standalone load balancer", option.Config.LoadBalancerDSRDispatch)
```
Comment (brb) [inline daemon/cmd/kube_proxy_replacement.go]: s/standalone load balancer/in lb-only mode/?

File: bpf/lib/lb.h [hunk 8]
```diff
@@ -1454,7 +1455,8 @@ static __always_inline int lb4_local(const void *map, struct __ctx_buff *ctx,
 #endif
 		tuple->daddr = backend->address;
 
-	return lb4_xlate(ctx, &new_daddr, &new_saddr, &saddr,
+	return skip_xlate ? CTX_ACT_OK :
+	       lb4_xlate(ctx, &new_daddr, &new_saddr, &saddr,
```
Comment (brb) [inline bpf/lib/lb.h:1459]: Maybe a comment saying that xlation will be handled by DSR code?

**Review by brb — APPROVED** [review]
Nice, LGTM! A few non-blocking comments.

Also:
1. I assume that you have tested it manually. However, how could we add tests to CI w/o extending bpf_lxc to decapsulate a request?
2. As a follow up it would be good to extend pkg/services/services.go to catch cases when backend port != service port in the IPIP dispatch mode (log a warning).

File: Documentation/cmdref/cilium-agent.md (hunk 6 above)
Comment (borkmann) [reply Documentation/cmdref/cilium-agent.md:38]: Given we need to handle this in Cilium in the mid-term as well, maybe it's not necessarily needed till next release.

File: daemon/cmd/kube_proxy_replacement.go (hunk 7 above)
Comment (borkmann) [reply daemon/cmd/kube_proxy_replacement.go]: Ok, will add once CI is through and green, then update and have smoke tests run.

**Review by pchaigno — APPROVED** [review]

**Comment by borkmann** [comment]
(all-green from CI side, only updated a help text in latest change)

**Comment by joestringer** [comment]
This was included in v1.9.2.

**Comment by kailunshi** [comment]
@joestringer i don't see this being backported to 1.9.*.  could you please double check?

**Comment by pchaigno** [comment]
I believe @kailunshi is correct. I don't see any of the commits in origin/v1.9 or any reference to the pull request in release notes for v1.9.x releases. The backports were dropped three times due to complexity issues. @borkmann is this still required for v1.9?

**Comment by borkmann** [comment]
> The backports were dropped three times due to complexity issues. @borkmann is this still required for v1.9?

It's not required, it was basically a fast-path optimization; 1.10 has it included that should suffice as an option.

**Pull request merged**